### DISCUSSION TITLE
반복 일정 기능을 위한 테이블 설계

### DISCUSSION THREAD

**Discussion by kmularise** [dailyge/dailyge-server#218] [Discussion]
## 📝 연관 PR
- [[DLG-365] 반복 일정 규칙 테이블을 설계하고 엔티티를 매핑한다.](https://github.com/dailyge/dailyge-server/pull/216)

<br/><br/>

## ❓ 작성 이유

반복 일정 관련해서 데이터베이스 테이블 설계 중 고민한 사항을 공유하고 팀원들과 회의한 내용을 기록하고자 작성했습니다.

<br/><br/>

## 📝 내용

반복 일정 관련해서 데이터베이스 테이블 설계 중 고려했던 사항은 다음과 같습니다.

1. 반복 일정의 설정 범위
2. 반복 일정에 대한 테이블 설계
<br/><br/>

## 1. 반복 일정의 설정 범위
반복 일정의 설정 범위 같은 경우에는 여러가지 케이스를 고려해야 했습니다.

### 반복 일정 설정 유형
1. 주간 반복
* 매주 특정 요일 (예: 월수금)
* 매주 단일 요일 (예: 월요일)
* 매주 휴일

2. 격주 반복
* 특정 주차의 요일 (예: 둘째 주 수요일)
* 특정 주차의 복수 요일 (예: 둘째 주 월,수)
* 홀수/짝수 주 반복

3. 월간 반복
* 매월 특정 일자 반복 (예: 매월 1일, 10일)
* 매월 특정 요일 반복

4. 격월 반복
* 짝수 월 반복
* 홀수 월 반복

5. 추가 설정 옵션
* 공휴일 제외 여부 설정

이러한 다양한 케이스들을 고려하여 반복 일정 설정의 유연성을 높일 수 있는 테이블 설계가 필요합니다. 사용자의 다양한 일정 관리 요구사항을 충족시킬 수 있도록 설계를 진행하고자 했습니다.

<br/><br/>

## 2. 반복 일정에 대한 테이블 설계

우선 반복 일정에 대해서는 기본적으로 tasks 테이블이 반복 일정 task_recurrences 테이블에 대한 id를 갖게 했습니다. 그 이유는 반복 일정을 삭제하거나 수정할 때, 관련된 task 데이터들을 빠르게 조회할 수 있다고 생각했기 때문입니다.

<br/><br/>

또한 반복 일정 테이블 설계에 반복 일정 규칙을 어떻게 반영할지에 대해 고민하였고, 크게 세가지 방안을 떠올렸습니다.

### 1. task_recurrences에 cron 표현식을 통해 반복 일정 규칙 나타내기

반복 일정을 cron 표현식을 통해 나타내고자 했습니다.
* 장점이라면 column 하나로 다양한 케이스들을 표현할 수 있습니다.
* 단점은 DB에 들어 있는 값을 직관적으로 알아보기 힘듭니다.
```mermaid
erDiagram
    tasks {
        BIGINT id PK "할 일 ID"
        BIGINT user_id "사용자 ID"
        VARCHAR(150) title "제목"
        VARCHAR(2500) content "내용"
        INT month "월"
        INT year "년"
        DATE date "날짜"
        ENUM status "할 일 상태(TODO,IN_PROGRESS,DONE)"
        VARCHAR(255) color "할 일 색상"
        BIGINT monthly_task_id "월간 일정"
        BIGINT task_recurrence_id FK "반복 일정 규칙 ID"
        TIMESTAMP created_at "생성일"
        BIGINT created_by "생성한 사람"
        TIMESTAMP last_modified_at "최종 수정일"
        BIGINT last_modified_by "최종 수정한 사람"
        BIT deleted "삭제 유무"
    }

    task_recurrences {
        BIGINT id PK "반복 일정 규칙 ID"
        VARCHAR(50) cron_expression "반복 일정 cron 표현식"
        VARCHAR(50) title "제목"
        VARCHAR(1500) content "내용"
        TIMESTAMP start_date "시작 날짜"
        TIMESTAMP end_date "끝나는 날짜"
        TIMESTAMP created_at "생성일"
        BIGINT created_by "생성한 사람"
        TIMESTAMP last_modified_at "최종 수정일"
        BIGINT last_modified_by "최종 수정한 사람"
        BIT deleted "삭제 유무"
    }

    task_recurrences ||--o{ tasks : "1:N"
```

### 2. task_recurrences에 복수의 column을 통해 반복 일정 규칙 나타내기
 task_recurrences 에 반복 일정 종류(주간, 월간) 컬럼과 반복 날짜 패턴, 공휴일 고려 유무를 통해 반복 일정 규칙을 나타냅니다.
 * 장점은 1.방안보다 DB에 있는 값을 직관적으로 알 수 있습니다.
 * 단점은  반복날짜패턴에 json이 들어가서 json 데이터 처리를 위한 추가 로직이 필요하게 됩니다.
```mermaid
erDiagram
    tasks {
        BIGINT id PK "할 일 ID"
        BIGINT user_id "사용자 ID"
        VARCHAR(150) title "제목"
        VARCHAR(2500) content "내용"
        INT month "월"
        INT year "년"
        DATE date "날짜"
        ENUM status "할 일 상태(TODO,IN_PROGRESS,DONE)"
        VARCHAR(255) color "할 일 색상"
        BIGINT monthly_task_id "월간 일정"
        BIGINT task_recurrence_id FK "반복 일정 규칙 ID"
        TIMESTAMP created_at "생성일"
        BIGINT created_by "생성한 사람"
        TIMESTAMP last_modified_at "최종 수정일"
        BIGINT last_modified_by "최종 수정한 사람"
        BIT deleted "삭제 유무"
    }

    task_recurrences {
        BIGINT id PK "반복 일정 규칙 ID"
        ENUM recurrence_type "반복 일정 종류(WEEKLY,MONTHLY)"
        JSON date_pattern "반복 날짜 패턴([1,2,3,4] or [1,31])"
        BIT consider_holiday "공휴일 고려 유무"
        VARCHAR(50) title "제목"
        VARCHAR(1500) content "내용"
        TIMESTAMP start_date "시작 날짜"
        TIMESTAMP end_date "끝나는 날짜"
        TIMESTAMP created_at "생성일"
        BIGINT created_by "생성한 사람"
        TIMESTAMP last_modified_at "최종 수정일"
        BIGINT last_modified_by "최종 수정한 사람"
        BIT deleted "삭제 유무"
    }

    task_recurrences ||--o{ tasks : "1:N"
```

3. 정규화를 위해 각 규칙을 나타내는 task_rule을 만들고 task_recurrences와 다대일로 매핑하기  

정규화를 통해 주간, 월간에 따라 반복 일정 규칙을 생성하는 테이블을 따로 만들고, 각각의 테이블이 task_recurrences_id를 갖고 있게 했습니다.
* 장점은 정규화가 된다는 점입니다.
* 단점은 테이블 간 조인이 더 많이 필요하게 됩니다.
```mermaid
erDiagram
    tasks {
        BIGINT id PK "할 일 ID"
        BIGINT user_id "사용자 ID"
        VARCHAR(150) title "제목"
        VARCHAR(2500) content "내용"
        INT month "월"
        INT year "년"
        DATE date "날짜"
        ENUM status "할 일 상태(TODO,IN_PROGRESS,DONE)"
        VARCHAR(255) color "할 일 색상"
        BIGINT monthly_task_id "월간 일정"
        BIGINT task_recurrence_id FK "반복 일정 규칙 ID"
        TIMESTAMP created_at "생성일"
        BIGINT created_by "생성한 사람"
        TIMESTAMP last_modified_at "최종 수정일"
        BIGINT last_modified_by "최종 수정한 사람"
        BIT deleted "삭제 유무"
    }

    task_recurrences {
        BIGINT id PK "반복 일정 규칙 ID"
        ENUM recurrence_type "반복 일정 종류(WEEKLY,MONTHLY)"
        BIT consider_holiday "공휴일 고려 유무"
        VARCHAR(50) title "제목"
        VARCHAR(1500) content "내용"
        TIMESTAMP start_date "시작 날짜"
        TIMESTAMP end_date "끝나는 날짜"
        TIMESTAMP created_at "생성일"
        BIGINT created_by "생성한 사람"
        TIMESTAMP last_modified_at "최종 수정일"
        BIGINT last_modified_by "최종 수정한 사람"
        BIT deleted "삭제 유무"
    }

    recurrence_weekly_patterns {
        BIGINT id PK "주간 패턴 ID"
        BIGINT task_recurrence_id FK "반복 일정 규칙 ID"
        INT day_of_week "요일 (1:월요일 ~ 7:일요일)"
        TIMESTAMP created_at "생성일"
        BIGINT created_by "생성한 사람"
        TIMESTAMP last_modified_at "최종 수정일"
        BIGINT last_modified_by "최종 수정한 사람"
        BIT deleted "삭제 유무"
    }

    recurrence_monthly_patterns {
        BIGINT id PK "월간 패턴 ID"
        BIGINT task_recurrence_id FK "반복 일정 규칙 ID"
        INT day_of_month "월의 날짜 (1-31)"
        TIMESTAMP created_at "생성일"
        BIGINT created_by "생성한 사람"
        TIMESTAMP last_modified_at "최종 수정일"
        BIGINT last_modified_by "최종 수정한 사람"
        BIT deleted "삭제 유무"
    }

    task_recurrences ||--o{ tasks : "1:N"
    task_recurrences ||--o{ recurrence_weekly_patterns : "1:N"
    task_recurrences ||--o{ recurrence_monthly_patterns : "1:N"
```

**Comment by devjun10** [comment]
좋네요. 이러면 반복 테이블을 통해 전체를 관리할 수 있을 것 같아요. 이건 저도 고려 못 했던 내용인데 하나 배워갑니다! 
> 추가로 항상 애플리케이션은 제약사항을 가지고 있어야 하는데, 반복 최대를 연간 단위로 잡으면 어떨까 싶어요. 즉, 반복 일정을 추가하면 올해 말, 2024.12.31일 까지만 등록하게 하는 옵션을 가지는거죠. 이 부분은 어떻게 생각하시나요?

**Reply by kmularise** [reply]
현재 상황에서 task_recurrences에 start_date,와 end_date로 반복일정이 적용되는 기간을 넣어두고 반복 일정 등록 시에 start_date와 end_date 차이가 1년 이상 나면 등록되지 못하게 제약사항을 걸어둘게요!

**Reply by devjun10** [reply]
좋네요 ㅎㅎ 충분할 것 같습니다.

**Comment by devjun10** [comment]
조금 걱정되는 부분은 Month 테이블도 생기면 너무 복잡해질 것 같단 생각이 드는데, 서비스는 최대한 간단할 수록 나중에 확장하기 쉽다고 생각이 들어요. 우선은 단순 반복을 하나 추가하고, 그 뒤에 다른 기능을 추가하면 어떨까 싶은데, 이 부분은 어떻게 생각하시나요?

**Reply by kmularise** [reply]
단순 반복이라고 하면, 어떤 반복일정인가요? 주간 반복일정을 말씀하시는 걸까요? 네 반복일정 규칙을 단순화하는 것도 좋을 거 같아요!

**Reply by devjun10** [reply]
주간반복 입니다!

**Reply by beatmeJY** [reply]
저도 동의합니다! 현재는 주간 반복(월, 화, 수, 목, 금, 토, 일)과 월간 반복(1일, 10일 등 지정)만 있으면 되지 않을까 하네요

**Reply by devjun10** [reply]
주간/월간을 동시에 체크했을 때 어떻게 할지도 고민해보시죠! 혹은 겹칠때도!

> 클라에서 막겠지만, 포스트맨과 같은 다른 툴을 이용해서 요청을 보냈을 때, 서버에서 어떻게 방어할지 등!

**Reply by kmularise** [reply]
그러면 월간 반복 또는 주간 반복만 선택하게끔 하고 (매달 하나 이상의 날짜 지정, 매주 하나 이상의 요일 지정) 동시 체크 관련해서는 방어 로직을 세워보도록 할게요!

**Reply by devjun10** [reply]
좋습니다. 그리고 비즈니스적으로 그냥 같은 시간대에 일정이 있다면 등록을 못하게 해버리는 것도 나쁘지 않을 것 같아요.

**Reply by kmularise** [reply]
아 일정 시작 시간도 고려하는 거군요 좋습니다!

**Reply by beatmeJY** [reply]
말씀해주신 3번 방법에서는 하나의 자식 테이블로 합쳐서 사용하는 것도 좋을 것 같아요 

task_recurrence_target (tinyint 컬럼)
  - 주간 반복: 0~6
  - 월간 반복: 1~30

**Reply by devjun10** [reply]
그냥 주간/월간 반복인지 ENUM으로 표현해도 되지 않을까요?

**Reply by devjun10** [reply]
날짜 계산은 로직으로 처리하고요. 혹시 주간/월간 이 기간을 기록해두려는 이유가 있을까요?

**Reply by kmularise** [reply]
좋은 의견입니다. 3번 방법을 사용하게 되면 기존에 weekly_patterns, monthly_patterns를 하나의 테이블로 합칠게요!

**Comment by devjun10** [comment]
마지막으로 cron을 **`어떤 식`** 으로 표현할지 나오면 좋을 것 같은데, 이 부분만 조금 더 설명해주시겠어요? 
> * * * * * * 이런 형태

**Reply by kmularise** [reply]
일반적인 크론 표현식에 의하면 이러한 형식의 String이 들어가게 될 거 같아요!
```
* * * * * *
│ │ │ │ │ │
│ │ │ │ │ └─ 요일 (0-7, 0 또는 7은 일요일)
│ │ │ │ └─── 월 (1-12)
│ │ │ └───── 일 (1-31)
│ │ └─────── 시 (0-23)
│ └───────── 분 (0-59)
└─────────── 초 (0-59)
```

만약 매월 1일이라고 하면 
```
0 0 0 1 * ? *
```
다음과 같은 크론표현식이 cron_expression이라는 컬럼 안에 들어가게 됩니다.

**Reply by devjun10** [reply]
사용자는 간단하게 주간/반복만 체크하면 저 값으로 바꿔준다는 거죠?

**Reply by kmularise** [reply]
네 맞습니다!

**Reply by devjun10** [reply]
그렇다면 크론표현식을 굳이 사용하지 않아도 괜찮지 않을까요? DB에 저장된 값을 읽어서 어떤 로직을 처리할텐데, ENUM으로 WEEKLY, MONTHLY 이정도만 해도 구분되지 않나? 하는 생각이 들어요. 

> 다른 케이스를 생각하시는게 있을까요?

**Reply by kmularise** [reply]
'매달 하나 이상의 날짜 지정, 매주 하나 이상의 요일 지정'과 같은 케이스만 고려한다면 크론표현식을 굳이 사용하지 않아도 될 것 같습니다!

**Comment by beatmeJY** [comment]
반복 일정을 삭제하게 되면 이미 지난 일정과, 지나지 않은 일정들이 존재하게 될텐데요.
이를 모두 삭제할지, 아니면 지나지 않은 일정만 삭제할 지에 대해서도 정하면 좋겠네요.

**Reply by kmularise** [reply]
반복 일정 삭제 시 이미 지난 일정은 그대로 두고 앞으로의 일정만 삭제할 것으로 생각했습니다! 이에 대해선 어떻게 생각하시나요?

**Reply by devjun10** [reply]
비즈니스 로직은 살짝 복잡해지겠지만 지난 일정은 굳이 삭제하지 않아도 될것 같아요. 👍

**Reply by beatmeJY** [reply]
좋습니다ㅎㅎ 그렇게 하시죠!

**Reply by devjun10** [reply]
Task쪽 비즈니스가 좀 들어가겠는데요? ㅋㅋㅋㅋ 재밌겠다 ✨✨
